### ISSUE TITLE
Automatically update node packages

### ISSUE THREAD

**Issue by ryelle** [WordPress/wordcamp.org#155]
Maybe we could use [renovate bot](https://renovatebot.com/), which we can configure to run once per sprint, and get one PR with a set of updates (rather than [a few a day](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pulls?q=is%3Apr+author%3Aapp%2Frenovate)). We could also enable/disable this per folder, with `includePaths` & `ignorePaths` (at least, I _think_ that's what these options control), so that only currently "active" projects get the updates.

When renovate makes PRs, it will tell you [what's changed in each package](https://github.com/woocommerce/woocommerce-admin/pull/2623) (pulling from the package's release, so this _can_ be hit or miss). Without automated testing this is a more manual process, but it should be a matter of building the projects and testing that everything works once the PR comes in.

We can even add this config in `.github/renovate.json` to keep the root uncluttered :)
 
The folders that could be updated (have `package.json`) are:

- `plugins/wcpt`
- `mu-plugins/blocks`
- `mu-plugins/theme-templates/templates/day-of-event`

Currently in the blocks folder we have 19 packages which are not at the latest version, most are minor updates but some are major - I've found in general it's easier to keep up with these when we have the automated help regularly reminding the whole team.

_I'm open to using another tool, renovate is just what I'm familiar with._

**Comment by iandunn** [comment]
What's the problem that you'd like to solve with this?

I can't think of any compelling benefits off the top of my head, except for the in the case of security updates (which we already get automated alerts about, and could enable automatic PRs if we wanted).

My main concern is taking our limited time away from making an impact with user-facing features, and adding more tedious busywork. If it'll save us time in the short-mid term, or offer some other compelling benefit, then I'm definitely open to considering it.

My secondary concern is that I'm leery using build tools liberally, [for the same reason as regex](https://blog.codinghorror.com/regular-expressions-now-you-have-two-problems/). That's not to say they don't have their place, but I feel like they're usually a pain to setup, break frequently and seemingly at random, are fragile across different dev environments, produce messy results (e.g., the title of #153, `npm install` overwriting coding styles in `package.json`, etc), etc.

IMO, they often add just as much complexity as they remove, if not more.

> We can even add this config in .github/renovate.json to keep the root uncluttered :)

That you for considering that! If we adopt this that would definitely be nice.

**Comment by coreymckrill** [comment]
I like the idea of automating the process of regularly auditing and updating our dependency packages, _and_, there are probably a few concerns that need to be addressed before we make a decision on this:

* How much would it cost? It looks like you get a free trial, but I didn't see an "always" free offering.
* What kind of permissions to our GitHub repo would we need to give it? Clearly Woo has considered this and deemed the service trustworthy, but it's still worth thinking about for our own codebase.
* What are the benefits of this service above and beyond what GitHub's built-in Dependabot gives us? Batching updates to once per sprint is one, but are there others?

**Comment by iandunn** [comment]
> What kind of permissions to our GitHub repo would we need to give it? 

IIRC GitHub's API permissions are very poorly designed, and it's not possible to grant read-only access to a single repo. If we connected an external app, it'd get read (and possibly write?) access to all repos in the organization, including private ones. I feel pretty strongly that that'd be a dealbreaker for security reasons. 

If we decide we want this functionality, is there a way to set it up without granting them any permissions to the organization?

**Comment by ryelle** [comment]
> What's the problem that you'd like to solve with this?

Our packages out pretty out of date, and keeping up with them somewhat regularly is easier than waiting months and needing to update all of them. Staying up with more-current versions could also make security updates easier when they need to happen, because it wouldn't need to be as big a jump  (instead of `random-package@2.0.3` to `3.1.6`, we could already be at `3.1.5`).

> My main concern is taking our limited time away from making an impact with user-facing features, and adding more tedious busywork.

That's fair, but aside from the initial cost of setup and that first big update, I don't think this would take much time out of the sprint (I also suggested enabling it only on select folders (like blocks) so this isn't overwhelming to start).

> The title of #153

That's a mess, I agree 😆  Renovate's a bit nicer, it would probably have done `Update dependency lodash to v4.17.13`.

> How much would it cost? It looks like you get a free trial, but I didn't see an "always" free offering.

Great question, I've only used it on orgs where it's already enabled (Automattic & WooCommerce)

> What are the benefits of this service above and beyond what GitHub's built-in Dependabot gives us? 

Not sure, I thought that was _just_ for security updates but maybe it's also configurable. A quick look says it can probably do all the same things, so that might be a better choice since it's already part of github.

**Comment by iandunn** [comment]
> keeping up with them somewhat regularly is easier than waiting months and needing to update all of them. 

I'm not sure that it is easier, but on a deeper level, I don't see why we need to install non-security updates in the first place. What tangible benefits do we get for doing that work?

> ...could also make security updates easier when they need to happen, because it wouldn't need to be as big a jump

I agree that it would make that discrete task easier, but my instinct is that it'd take more time if you look at it holistically. In fact, jumping over several versions seems like it might actually _save_ time overall. 

1) In the case where there aren't any back-compat breaks, then updating to the intermediary versions takes extra time to test/merge/deploy/etc each individual update, but doesn't save any time in the future.

2) In the case where there's a single back-compat break, it seems like it'd take the same amount of time to update, whether we do it now or in the future.

3) In the case where there're multiple back-compat breaks, it seems like it'd either save time (b/c you'd have to refactor anyway and you can just refactor against the latest version), or take the same amount of time (for the same reason as above).

Even if the above aren't accurate, it seems like we'd be _guaranteeing_ ourselves extra work today in order to _maybe_ save time tomorrow. Not all of those packages contain vulnerabilities in the branches that we run, so we'd be updating some of them unnecessarily. In fact, upgrading to new major releases will probably contain new features we don't need, which could just as easily introduce a vulnerability that doesn't exist in the older branch.

Doing the updates every sprint would also introduce additional context switching and overhead, and I worry that would distract me from focusing on impact even more than I already am. Those extras wouldn't exist if we batched the updates together at the point in the future where there's a tangible reason to update (e.g., security, or a bug that organizers are significantly affected by).

**Comment by ryelle** [comment]
> What tangible benefits do we get for doing that work?

Whatever the updates are bringing- improved accessibility, performance, edge-case bugfixes that we maybe haven't run into yet but then don't have to worry about… the same reasons we want to keep plugins etc updated.

> In fact, jumping over several versions seems like it might actually save time overall.

Having debugged multiple outdated projects where the build process itself has broken down, I definitely disagree with this. I see where you're going with each part individually, but the issues really compound once you have multiple things outdated, because then you need to update all of them at once, possibly update configuration, build tools, etc

> Doing the updates every sprint would also introduce additional context switching and overhead, and I worry that would distract me from focusing on impact even more than I already am.

I haven't been the maintenance person on this team yet so I'm not all sure what goes into that job, but that's the role we had doing this on my last team– you're already at a shallower context triaging PRs and issues, taking an hour to run through the updates didn't seem like a huge impact in practice. I understand this might be different for different devs & the maintenance role might be different here.

**Comment by coreymckrill** [comment]
@ryelle I agree with you about the value of doing this, and I defer to your experience that keeping up with this regularly has worthwhile benefits and is more efficient timewise than letting things get stale.

I'm not sold on renovate bot in particular though, for reasons mentioned above. Are there other similar services available that we could compare it to? Or, is there a relatively painless way we could do this ourselves, auditing our packages periodically and getting a list of things that should be updated?

**Comment by ryelle** [comment]
> Are there other similar services available that we could compare it to? 

I think we could look into dependabot. I installed the "Dependabot Preview" app to a personal repo of mine, and this is what it asked for. Note that it's written/owned by github, so they already have this access theoretically. It's only installed as a separate app because it previously was separate (was acquired by github)

![Screen Shot 2019-07-18 at 12 41 37 PM](https://user-images.githubusercontent.com/541093/61475742-7fa1bb80-a959-11e9-8de5-f41f75c2ad9c.png)

Looks like we can also set this to do weekly or monthly updates, but otherwise not all the advanced config of renovate. I haven't gotten a PR from it yet, so I'm not sure if it shows all the same info as renovate.

> Or, is there a relatively painless way we could do this ourselves, auditing our packages periodically and getting a list of things that should be updated?

It's possible to look at `npm outdated`, but then you need to check the new versions manually - the thing I liked best about renovate was the Release Notes section where it just tells you, right there, what's changed:

![Screen Shot 2019-07-18 at 12 37 41 PM](https://user-images.githubusercontent.com/541093/61475481-e1155a80-a958-11e9-9436-d953ca17e975.png)

I think doing this manually would be a time-sink outweighing the benefits.

**Comment by ryelle** [comment]
Looks like dependabot does include the changelog + release notes (they're the same in this PR though) https://github.com/ryelle/rmb-recipe-block/pull/95

This "compatibility score" is pretty interesting: https://dependabot.com/compatibility-score/?dependency-name=css-loader&package-manager=npm_and_yarn&previous-version=3.0.0&new-version=3.1.0

**Comment by coreymckrill** [comment]
If the feature set is comparable, and Dependabot is free (which it says it is, with no "premium" version mentioned, since it's now part of GitHub), that seems like the best choice to me.

I wonder if all of the rest of the Dependabot features will be integrated right into GitHub, so that we eventually wouldn't have to give permissions to a separate app. I'd feel better about that, even though it's technically owned by GitHub.

**Comment by coreymckrill** [comment]
These points from the Dependabot blog [announcement](https://dependabot.com/blog/hello-github/) seem to indicate that eventually everything will be integrated:

> * We're integrating Dependabot directly into GitHub, starting with security fix PRs 👮‍♂️
> * You can still install Dependabot from the GitHub Marketplace whilst we integrate it into GitHub, but it's now free of charge 🎁
> ...
> * In time, you'll be able to configure Dependabot within GitHub, so you'll no longer need the Dependabot dashboard.

**Comment by ryelle** [comment]
Yeah, eventually - but with no timeline that I saw, so I think the "external" app is still needed if we want this soon.

**Comment by iandunn** [comment]
> improved accessibility, performance, edge-case bugfixes that we maybe haven't run into yet but then don't have to worry about… 

This is subjective, but I feel like the updates are probably just as likely to _cause_ performance issues and a11y/regular bugs than they are to _solve_ them, because most software tends to add bloat that we don't need and will never use. Any bugs that might exist in those outdated packages are so small that they haven't been noticed by us or reported by anyone else, so it doesn't seem worth it to me to spend a bunch of time updating them on the off chance that it _might_ have a benefit. 

> the same reasons we want to keep plugins etc updated.

That's a fair point, but I think there are some fundamental differences in this context, like how  plugins tend to be user-facing, and are therefore more more likely to add features that people can use automatically; whereas libraries are developer-facing, and often require extra work on our part to take advantage of a new feature.

`<rant>`
To be honest, though, if there was a way to get reliable automated notifications for WP plugin security vulnerabilities, and WP itself wasn't so obnoxious about badgering me to update something every other day, then I'd be thrilled to postpone plugin/themes updates too, until a compelling, tangible reason to update came along.

In general, I feel like most software updates are a huge waste of time for the user, and rarely add anything of meaningful value. Most of the software we run is already [complete](https://engineering.hmn.md/how-we-work/philosophy/completion/) for our purposes.

In my ideal world, security updates would be installed automatically, and all other updates would be silently ignored until the user encounters a reason compelling enough to make them wonder if an update is available that would benefit them, and then go search for it. I feel like it's obnoxious, distracting, and counter-productive that I get interrupted every couple of days by some random piece of software that thinks it's pointless update is more important than [my ability to be intentional about what I focus on](https://uxdesign.cc/introducing-mercury-os-f4de45a04289?gi=ff68b70f24b8).
`</rant>`

At the end of the day, we have very limited time, and there are much more impactful things we could be spending it on. All of the little things add up and consume a ton of time, and I find that it makes it difficult for me to focus on things that actually matter.

We've been just fine for years without doing this, so I don't think it makes sense to add even more low-impact work to our plates when we already spend so little time on things that really move the needle.

----

Having said that, we've also considered automating plugin/theme updates before, and I'd be open to doing that here as well (for non-breaking updates), if there's a way to do it without dev intervention/testing/etc, and we manually deploy the changes (so any problems are more likely to show up when someone is around to fix them).

And of course, if you're both still in favor this, then I can definitely _disagree and commit_.

**Comment by coreymckrill** [comment]
If we can use Dependabot with only this repo (and not give the app access to the whole organization), I'm in favor of trying it out for some amount of time, say 3-6 months. After that we can evaluate how noisy it is, and whether the updates it provides to us have been valuable and worth our time.
